### DISCUSSION TITLE
Petition to rename Scripts.

### DISCUSSION THREAD

**Discussion by kovitikus** [evennia/evennia#2385] [Evennia Engine Development]
Given the nature of scripts, not actually being scripts but so much more, I'd like to petition for them to be renamed for 1.0 to widgets. If there were ever a time to do this, it would be now.

I think it will break entire projects because the typeclass path will change, but YOLO.

***

There was a lot of discussion that happened in the live chat about this topic, so I'll link that here as well. It is linked to the message I sent that sparked the whole discussion on scripts and their name and usage.

https://irclogs.jackgrigg.com/irc.freenode.net/evennia/2021-04-19#i_4715581

***

Ok, I've updated the name of the discussion post. We've ruled out the notion of Widget, due to conflicts with Django.

We've landed on a few reasonable suggestions:

```
Script
Component
Resource
OOCObject
Gizmo
```

You can vote on the name you like best here on this straw poll. The poll only allows one choice per IP address, but I implore you not to abuse the poll and vote twice. It's going to be most helpful to get proper feedback on this.

http://www.strawpoll.me/42967195

**Comment by taelspinner** [comment]
I'm firmly behind this motion and adding my name. While it's a minor nuisance, I think it would help with the onboarding process if they were given a name that's more fitting to their actual purpose and flexibility.

And if those people didn't want their projects broken, they shouldn't have been developing open source, now should they? :P

**Comment by davewiththenicehat** [comment]
Would it be possible to make a widget handler that forwards to scripts?
It's been a good long while since I looked at the scripts setup.

**Reply by taelspinner** [reply]
I imagine you could quite easily add in a backwards-compatible alias similar to how `list_to_string` works for `iter_to_string`.

**Comment by PeaceBrigade** [comment]
I agree that "script" may not be the proper term for what we currently use scripts for. I think if we're looking for new names more should be considered though especially if onboarding is the intention. Almost all technology and programming courses of the last 5-10 years, as well as some of the biggest and fastest rising software ecosystems (namely Mobile and Gaming), all already use "Widget" as a solidified term for specific use-cases. 

I'm not saying that students are the target audience, hell, I think the target audience for MU* in general are ancient grognards. "Widget" is a term already cemented in educational teachings and courses in countless programming courses. It is synonymous with being a user interface component, something that text-based games run kind of counter to.

I'm not sure what to propose in its stead. Addons are extensions of other objects, and that doesn't seem to fit with the use of evennia "scripts."
 Object is taken, perhaps Component? Component seems maybe the most apt description, since scripts are just being currently used as whatever in the wild-west of Evennia.

I'll think on it some more but it's actually an interesting question, I'll brainstorm with my students tonight, there might be some magical unicorn term that just isn't coming to mind that better fits, but renaming something from one wrong thing to another doesn't seem like a good idea. Ultimately I am sure it is up to the maintainer, but I would like to see some serious consideration given if we're going to change the terminology and break all previous projects.

**Reply by taelspinner** [reply]
> "Widget" is a term already cemented in educational teachings and courses in countless programming courses. It is synonymous with being a user interface component, something that text-based games run kind of counter to.

Personally, I think this is precisely why it's absolutely fine to use here. Outside of those programming courses for mobile developers, "widget" is not a very commonly used software term, and it's never used in the context of a text-based multiplayer telnet game. So I doubt there's going to be much confusion among the general public, particularly those who don't develop mobile games or apps, and those who do have a familiarity with it will quickly realize in context that it's borrowing the term from a broader, non-industry sense.

I'm open to Component as an alternative. It's really the only term that comes close to the connotation of widget, which did, before Web 2.0 and smartphones, really basically just mean "something that does whatever the hell it does."

**Reply by PeaceBrigade** [reply]
Huh? Widget is one of the most commonly used software terms as it is one of the most widely understood terms by users. Most users might not know what most programming terms or components are, but Widget is I think an exception. There is probably no one under the age of perhaps .... 25-30 who doesn't instantly identify it with a specific concept.

I can see an argument for the difference between what a back-end programmer refers something to, and a user, but most of the people I believe taking programming courses in the last decade have a defined notion of what that specific term applies to, so I think we should avoid it. I think it's also extremely important to not shoehorn terms known in the industry into projects while maintaining that we're using them in a non-industry sense. It's best to avoid confusion especially for those who might be coming from a newer education where certain facets of what they have been taught needs to be ignored because framework developers have decided on non-conventional names for things.

I think that Component might be good as well. It seems to fit decently, and the only issue I personally have with it is that I spend 60+ hours a week coding, and Component is such a long word. 

Seeing as the other normal terms I might apply in this situation like addon, plugin, module, object, etc are either protected in python, or run into the same definition problems as script or widget in my eyes, I think Component is the most neutral and flexible representation of what we're probably trying to do with the object in question.

**Reply by taelspinner** [reply]
By any chance, do you teach mobile app/game development?

**Reply by kovitikus** [reply]
> Widget is one of the most commonly used software terms as it is one of the most widely understood terms by users. Most users might not know what most programming terms or components are, but Widget is I think an exception. There is probably no one under the age of perhaps .... 25-30 who doesn't instantly identify it with a specific concept.

A lot of assumptions here about what others consider to be a widget.

Many common terms are rehashed to fit the specific nature of a coding language or project. It is very appropriate to appropriate the term widget to represent an anomalous Evennia object. The term itself describes something that isn't easily defined and that's the nature of scripts in Evennia.

**Reply by Griatch** [reply]
I must agree with @PeaceBrigade here on the use of 'widget'. Apart from widget being a well established term for an UI element in general, [widget](https://docs.djangoproject.com/en/3.2/ref/forms/widgets/) is the term for a graphical element also in Django, which Evennia uses.

Component ... not sure. Maybe? 

Edit: I see @strikaco already pointed out the Django use.

**Reply by taelspinner** [reply]
The Django use is a pretty good reason not to name it that IMO; type collision like that should be avoided. Component is a more esoteric conflict; but that doesn't leave you with a lot of other choices. Someone in the Discord suggested "Resources"?

**Comment by strikaco** [comment]
"Widget" is a common web-development term, and [one that is actively used by Django itself in a different context.](https://docs.djangoproject.com/en/3.1/ref/forms/widgets/)

"Component" is also a confusing choice of term; the entire _Object_ prototyping system is a rebranding of entity-_component_ modeling...and "component" in this context has so little to do with _Scripts_, the spawner itself won't even create them.

"Scripts" is pretty free-form; you can do literally anything with them in _every_ context. Why the need for a rename?

**Reply by CloudKeeper** [reply]
I share strikaco's view.

**Reply by kovitikus** [reply]
My request for a rename is because the word script doesn't represent the free form nature of them. It's my opinion that scripts are just one layer of what they represent and not even something I often use them for.

I did not know that widget was used in django, but it makes sense. It is a different context. When I say django widget, you know that means specific to that concept, but if I say Evennia widget it would mean the thing that is currently a script.

**Reply by taelspinner** [reply]
> "Scripts" is pretty free-form; you can do literally anything with them in _every_ context. Why the need for a rename?

Well, because it's a confusing choice of term, obviously; scripts generally refer to linear sets of instructions that are attached to an object (this is how other MUDs use the term, too, worsening the confusion), whereas in Evennia they are discrete objects that don't need to include instructions at all, and are commonly used as incorporeal objects that don't have a place in the gameworld (for instance, for storing clothing configurations). New developers to Evennia from almost any background and stripe need to be corrected on their misconceptions about Scripts because "script" is a term that does not refer to what Scripts in Evennia do. I mean, Python is a *scripting language* -- what exactly does scripting have in common with what Scripts are used for? Griatch himself has mentioned before that this is a concern for him.

None of these terms are perfect, but I believe that Widgets or Components would *better* describe them than Scripts, and of the two, Widgets is more succinct.

**Comment by PeaceBrigade** [comment]
I have no problem and prefer them to remain scripts, I'm just saying that
if for some reason people wail loud enough to get it changed, I don't think
widget is the right term in any context

**Comment by PeaceBrigade** [comment]
I think this in itself is an argument why it shouldn't be used. Evennia is
so tightly integrated with Django that we shouldn't be using Django
specific terminology with a "but not in this case" caveat attached.

**Reply by kovitikus** [reply]
The same word can have a different meaning based on context. Let me use your own argument against you here.

This document outlines the definition of widgets, as is used in web applications.

https://www.w3.org/TR/widgets/

At the very beginning of this document they specify:
> Unlike traditional user interface widgets (e.g., buttons, input boxes, toolbars, etc.), widgets as specified in this document are full-fledged client-side applications that are authored using technologies such as HTML and then packaged for distribution.

Do you see how they had to make a distinction about which widgets they were referring to? You know why? Because the term widget is already used for other things. In this case, it is specified as a GUI widget. This is the nature of how terms are defined, by their context and use.

**Comment by PeaceBrigade** [comment]
I agree, I don't think that there needs to be a rename in the first place,
but I guess some people are taking issue with it. I would rather not have
any changes, as it would completely destroy any backwards compatibility
during an upgrade. (Though 1.0 might do this anyways)

**Comment by Griatch** [comment]
While I agree that _Scripts_ is not really the best name today (with the change in their use over time), it's not so obvious what one would rename them to. 

If they were to change I'd me more tempted to call them something like `OOCObjects` or something like that ...

**Reply by taelspinner** [reply]
`AetherObjects` 😃

**Reply by Griatch** [reply]
`KitchenSink`

**Reply by taelspinner** [reply]
`Thing`

**Reply by kovitikus** [reply]
@Griatch Isn't a command an object and is technically OOC?

I'm willing to throw in `gizmo` here, as it is similar to widget but completely unique. It is also more likely to stand out and be easily identified.

I'm also trying to keep the name short, because it's easier to type and reference.

**Reply by tsal** [reply]
`Adapters`

we all need more Adapters.

**Reply by volundmush** [reply]
Strangely enough, 'Script' is also what a few other game engines call it.

**Comment by TehomCD** [comment]
I would strongly, strongly prefer to have some alias added rather than do a model rename, with all the messiness that causes (every Attribute using a natural key immediately becomes broken, every import immediately breaks, etc). I'm indifferent to what aliases they might have, I just don't want to do a repeat of the Player -> Account process.

**Comment by TehomCD** [comment]
Thinking about this a bit more, might it be more natural to add a model to the object app that's 'OOCObject'? You could make it much more lightweight than either Scripts or Objects due to a reduced number of fields while being more explicit about its purpose. That would then leave Scripts more focused on their original purpose - a database wrapper for looping calls.

**Reply by kovitikus** [reply]
Please, god, don't name it OOCObject. That's such a terrible name.

**Reply by taelspinner** [reply]
`ThinObject`

**Reply by kovitikus** [reply]
`SupercalifragilisticexpialidociousObject`

**Reply by TehomCD** [reply]
Widget, Thingy, ObjectWithNoInGamePresence, NotAnOOCObject, whatever. The point is that if we're trying to repurpose scripts into something else, it makes more sense to add another model to suit that specific purpose rather than pound a square peg into a round hole.

**Reply by taelspinner** [reply]
Well, that's the thing, though. We aren't repurposing Scripts. We're trying to come up with a better term for them, because their purpose has (presumably?) broadened enough in the current build that the term no longer really fits, in my opinion.

**Reply by kovitikus** [reply]
@Griatch is rewriting scripts as we speak. I don't know the nature of the changes he's making, but perhaps he could give a TLDR.

**Reply by TehomCD** [reply]
Has it, though? Scripts are designed specifically to be wrappers for recurring events. People using them as 'ooc objects' without a looping call is functionally no different than using objects with a null location aside from them being accessed with the `script` command. I think it makes much more sense to make bespoke entities to fill this specific requirement of a locationless object that holds data/behavior rather than bundle them into scripts.

**Reply by kovitikus** [reply]
https://www.evennia.com/docs/latest/Scripts.html

***

Scripts are the out-of-character siblings to the in-character Objects. Scripts are so flexible that the “Script” is a bit limiting

We had to pick something to name them after all. Other possible names (depending on what you’d use them for) would be OOBObjects, StorageContainers or TimerObjects.

Scripts can be used for many different things in Evennia:

They can attach to Objects to influence them in various ways - or exist independently of any one in-game entity (so-called Global Scripts).

They can work as timers and tickers - anything that may change with Time. But they can also have no time dependence at all. Note though that if all you want is just to have an object method called repeatedly, you should consider using the TickerHandler which is more limited but is specialized on just this task.

They can describe State changes. A Script is an excellent platform for hosting a persistent, but unique system handler. For example, a Script could be used as the base to track the state of a turn-based combat system. Since Scripts can also operate on a timer they can also update themselves regularly to perform various actions.

They can act as data stores for storing game data persistently in the database (thanks to its ability to have Attributes).

They can be used as OOC stores for sharing data between groups of objects, for example for tracking the turns in a turn-based combat system or barter exchange.

**Reply by taelspinner** [reply]
***Is*** that what they're designed for, though? From the Evennia docs for Scripts:

> Scripts are the out-of-character siblings to the in-character Objects. Scripts are so flexible that the “Script” is a bit limiting
> - we had to pick something to name them after all. Other possible names (depending on what you’d use them for) would be OOBObjects, StorageContainers or TimerObjects.
>
> Scripts can be used for many different things in Evennia:
> - They can attach to Objects to influence them in various ways - or exist independently of any one in-game entity (so-called Global Scripts).
> - They can work as timers and tickers - anything that may change with Time. But they can also have no time dependence at all. Note though that if all you want is just to have an object method called repeatedly, you should consider using the TickerHandler which is more limited but is specialized on just this task.
> - They can describe State changes. A Script is an excellent platform for hosting a persistent, but unique system handler. For example, a Script could be used as the base to track the state of a turn-based combat system. Since Scripts can also operate on a timer they can also update themselves regularly to perform various actions.
> - They can act as data stores for storing game data persistently in the database (thanks to its ability to have Attributes).
> - They can be used as OOC stores for sharing data between groups of objects, for example for tracking the turns in a turn-based combat system or barter exchange.

**Reply by PeaceBrigade** [reply]
We have out of band messages and stuff. Let's not miss this opportunity for ... wait for it... OOBjects :D
